### ISSUE TITLE
Small possible divergence with original DQN paper

### ISSUE THREAD

**Issue by alcinos** [openai/baselines#41]
Hi,

I noticed that there might be a slight difference between this implementation of the network and the original one by DeepMind. Maybe this is a known fact, but I didn't see it mentionned anywhere, and as this implementation seems to try to be as close as possible to the original one, I thought it'd be worth it to point it out.

It boils down to the fact that this implementation uses the default padding from TensorFlow, which is 'VALID', whereas DeepMind didn't document any padding on their convolutional layers.
If we refer to the Torch implementation they released, we can conclude that they used the default padding of Torch (which is 0 in the SpatialConvolution module), except for the first layer, where they used padding=1

After the convolutions, the image sizes are quite different: 7x7 for (py)Torch and 11x11 for TensorFlow (40% difference). As such, the input size of the first linear layer diverge (3136 vs 7744)

I'm not sure that makes a huge difference (be it positive or negative) in the outcome, but experience has proved that devil's in the details when it comes to deep architectures.

What do you guys think ?

**Comment by awilliamson** [comment]
@alcinos I've seen things like this before when trying to recreate models from literature. From a CNN stance, it has mattered quite a bit. It depends what your application is. In terms of academia, I'd say try and get it as close as possible for comparisons.
Would you be able to modify the network, and run two side-by-side experiments showing if there is a significant difference between this version and the original DeepMind paper?
These little 'not mentioned' things such as padding are sometimes the crux of a problem when re-implementing having had a similar issue when fixing AlexNet.

**Comment by alcinos** [comment]
@awilliamson Unfortunately running experiments two test the difference is beyond my computational reach at the moment. To be significant, it has to be several games and on several seeds, and I don't have access to powerful gpus.
But if others are willing to try, I'd be glad to hear about the outcome.

**Comment by awilliamson** [comment]
@alcinos If you want to throw some code together, I'd be able to run it on a dual 1080 research machine. I simply don't have the development time currently, but I'm happy to run some experiments overnight to obtain the outcome.

**Comment by alcinos** [comment]
@awilliamson if we ignore DeepMind's padding of 1 on the first convolution, one easy way to make it similar to the original code is to add `padding = 'VALID'` in the parameters of all convolutions in model.py. That way, the shape of the image coming out of the last convolution is as expected.

For the padding of 1 in the first convolution, I don't know tf enough to figure out what's the correct way to do that.

**Comment by siemanko** [comment]
@alcinos Thanks for pointing that out. The goal of this project is to produce competitive results with an algorithm as close as possible to original paper. There are few more minor differences, e.g. learning rate schedule.
